hack/generate-ta-tasks.sh

```diff
@@ -1,29 +1,23 @@
 #!/usr/bin/env bash
 
 set -o errexit
+set -o errtrace
 set -o nounset
 set -o pipefail
 set -o posix
 
 shopt -s globstar nullglob
 
+command -v go &> /dev/null || { echo Please install golang to run this tool; exit 1; }
+[[ "$(go env GOVERSION)" == @(go1|go1.[1-9]+(|.*|rc*|beta*)|go1.1[0-9]+(|.*|rc*|beta*)|go1.20*) ]] && { echo Please install golang 1.21.0 or newer; exit 1; }
+
 HACK_DIR="$(realpath "$(dirname "${BASH_SOURCE[0]}")")"
 ROOT_DIR="$(git rev-parse --show-toplevel)"
 TASK_DIR="$(realpath "${ROOT_DIR}/task")"
 
-if ! command -v tash &> /dev/null; then
-  echo INFO: tash command is not available will download and use the latest version
-  tash_dir="$(mktemp -d)"
-  trap 'rm -rf ${tash_dir}' EXIT
-  tash_url=https://github.com/enterprise-contract/hacks/releases/download/latest/tash
-  echo INFO: downloading from ${tash_url} to "${tash_dir}"
-  curl --no-progress-meter --location --output "${tash_dir}/tash" "${tash_url}"
-  echo INFO: SHA256: "$(sha256sum "${tash_dir}/tash")"
-  chmod +x "${tash_dir}/tash"
-  tash() {
-    "${tash_dir}/tash" "$@"
-  }
-fi
+tash() {
+  go run -C "${ROOT_DIR}/ta-generator/" github.com/konflux-ci/build-definitions/ta-generator "$@"
```

Comment:
This seems significantly slower than pre-compiling

```bash
time GOTOOLCHAIN=auto GOSUMDB=sum.golang.org hack/generate-ta-tasks.sh
GOTOOLCHAIN=auto GOSUMDB=sum.golang.org hack/generate-ta-tasks.sh  7.21s user 1.48s system 168% cpu 5.168 total
```

vs. compiling once per script run

```bash
time GOTOOLCHAIN=auto GOSUMDB=sum.golang.org hack/generate-ta-tasks.sh
GOTOOLCHAIN=auto GOSUMDB=sum.golang.org hack/generate-ta-tasks.sh  1.61s user 0.37s system 166% cpu 1.192 total
```

Reply:
Yeah, do we care? I guess we might?

Reply:
Not in CI, but locally it can get annoying if you're trying to fix a recipe and regenerating files often

Reply:
Huh, I see the opposite (ran a couple of times, but getting consistent results):

`go run` version:
```
$ time hack/generate-ta-tasks.sh
hack/generate-ta-tasks.sh  12.54s user 2.80s system 170% cpu 8.990 total

$ time hack/generate-ta-tasks.sh
hack/generate-ta-tasks.sh  12.60s user 2.76s system 169% cpu 9.040 total
```

`go build` once and execute version:
```
$ time hack/generate-ta-tasks.sh
hack/generate-ta-tasks.sh  19.23s user 3.26s system 179% cpu 12.553 total

$ time hack/generate-ta-tasks.sh
hack/generate-ta-tasks.sh  19.24s user 3.13s system 181% cpu 12.347 total
```

I'll push the build & execute version, perhaps you can spot something

Reply:
Hmm, it seems to rebuild tash every time, but I can't tell why :thinking: 

```diff
diff --git a/hack/generate-ta-tasks.sh b/hack/generate-ta-tasks.sh
index 8268155..2762c02 100755
--- a/hack/generate-ta-tasks.sh
+++ b/hack/generate-ta-tasks.sh
@@ -16,6 +16,7 @@ ROOT_DIR="$(git rev-parse --show-toplevel)"
 TASK_DIR="$(realpath "${ROOT_DIR}/task")"
 
 tash() {
+  echo building tash >&2
   tashbin="$(mktemp --dry-run)"
   GOTOOLCHAIN=auto go build -C "${ROOT_DIR}/ta-generator/" -o "${tashbin}"
   trap 'rm "${tashbin}"' EXIT
```

```shell
$ GOTOOLCHAIN=auto GOSUMDB=sum.golang.org hack/generate-ta-tasks.sh
building tash
building tash
building tash
building tash
building tash
building tash
building tash
building tash
```

Reply:
It's my shell scripting, I'll figure this out

Reply:
So, fun, `tash` is executed in a subshell on the `sponge=$(tash...` line, so the redefinition of `tash` function is happening in the subshell and it is not propagated to the parent shell, so next time it is invoked the original non-memoized version is called. I've simplified it now, no need to be fancy here...

Reply:
:exploding_head: yep, bash is fun